### PR TITLE
[docs] Add some string-ascii and string-utf8 types to the docs

### PR DESCRIPTION
**Pull request by gregorycoppola**
Fixes blockstack/stacks-blockchain#2650

## Changes
This PR changes the documents for the following functions: map, filter, fold, concat, as-max-len?, len, element-at, index-of.

## Background
From blockstack/stacks-blockchain#2650:

"When string-ascii and string-utf8 were introduced, many functions documentation in the Clarity reference were not modified to include references to these new types when applicable."

In fact, some times the examples of valid usage already include the strings, but the function documentation does not say this is allowed, e.g., "len", "element-at".

## Testing
1) I ran cargo test vm::docs

This should mostly just test that the new file compiles.

2) I tested the new documentations that I added by creating a contract with the following print statements.

   ;;; tested for the PR
   ;;; map
   (print (map my-and 0x010203 0x010203))
   (print (map add-one-char u"hello world"))
   (print (map add-one-ascii "hello world"))
   ;;; filter
   (print (filter is-three 0x01020304030201))
   (print (filter is-vowel u"hello world"))
   (print (filter is-vowel-ascii "hello world"))
   ;;; concat
   (print (concat "concat1" "concat2"))
   (print (concat "concat1" "concat2"))
   ;;; len
   (print (len "hello world"))
   (print (len u"hello world"))
   ;;; element at
   (print (element-at u"hello world" u2))
   (print (element-at "hello world" u2))
   ;;; index of
   (print (index-of u"hello world" u"l"))
   (print (index-of "hello world" "l"))

### CODE REVIEW

**Review by kantai — CHANGES_REQUESTED** [review]
These changes look great, thanks @gregorycoppola!

The only thing I would add is some more examples for the changed docs, for example, adding like

```(define-private (buff-filter (a (buff 1))) (not (is-eq a 0x00)))
(filter buff-filter 0x00110022003300440055) ;; Returns 0x1122334455
```

And some similar ones for `string-ascii` (we probably don't need examples for both `string-ascii` and `string-utf8`).

**Comment by gregorycoppola** [comment]
Now "map", "filter", "fold", "concat", "as-max-len", "len", "element-at", and "index-of", should each have "string" and "buff" examples.

**Comment by gregorycoppola** [comment]
> These changes look great, thanks @gregorycoppola!
> 
> The only thing I would add is some more examples for the changed docs, for example, adding like
> 
> ```
> (filter buff-filter 0x00110022003300440055) ;; Returns 0x1122334455
> ```
> 
> And some similar ones for `string-ascii` (we probably don't need examples for both `string-ascii` and `string-utf8`).

**Comment by gregorycoppola** [comment]
I closed the comment.. not sure if I was supposed to close it or let you close it, but let's see.

**Comment by psq** [comment]
> I closed the comment.. not sure if I was supposed to close it or let you close it, but let's see.

that closed the whole PR, not what you wanted, I'm guessing, so I reopened so this does not get lost. Sometimes, for inline comments, you can mark as `resolved` but not in this case.

Nice to see all these extra examples, thank you for doing this!

**Comment by gregorycoppola** [comment]
@psq Thanks for reopening :)

**Review by kantai — APPROVED** [review]
LGTM, thanks @gregorycoppola!

**Comment by gregorycoppola** [comment]
Thanks @kantai .. added @pavitthrap for second review.

File: src/vm/docs/mod.rs
Comment (pavitthrap) [inline src/vm/docs/mod.rs:603]: Same as above - do we want the or syntax?

File: src/vm/docs/mod.rs
Comment (jcnelson) [reply src/vm/docs/mod.rs:534]: Seconded.  I see that `|` is used in other places in this diff to represent a choice between types.  Can the documentation be consistent?

File: src/vm/docs/mod.rs
Comment (jcnelson) [inline src/vm/docs/mod.rs:592]: Can you add an example here for `buff`?

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:469]: It really does have to be (buff 1). That is a '(list A)' is made up of objects of type 'A', while a 'buff' is made up of objects of type '(buff 1)'

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:501]: done

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:516]: I updated each usage to be "(where a/each sequence is either `buff`, `list`, `string-ascii` or `string-utf8`)"

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:592]: I think there already is one:
`(element-at 0xfb01 u1) ;; Returns (some 0x01)`

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:489]: I removed them.. Is there really a "right way to do this"?

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:534]: The '\n instead of |' was an idea that @kantai and I came up with to avoid overly long lines. The idea was that '\n' would have the same semantics a '|'. But, there are definitely a confusing number of variations.

Also, I just tested how some of these changes look using the 'docs' repo. It seems that the html generator doesn't respect the newline anyway. It seems the line just runs on, so this probably isn't a workable idea.

https://paste.pics/fe67b8ecfe29e5ba945f9728f5b6b349

On the other hand, even if we put the '|' in the line, it will still be really hard to read, if the line still just runs on.

@jcnelson how would you want to handle this?

What do you think of, e.g., the following for `fold`?

`Function((buff 1)|A|(string-ascii 1)|(string-utf8 1), B) -> B, buff|(list A)|string-ascii|string-utf8, B`

To me, it seems a little confusing, but this might be the most consistent way to handle this case.

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:603]: see discussion above

**Comment by gregorycoppola** [comment]
Thanks for the reviews @pavitthrap and @jcnelson.

I addressed and resolved some comments, but left some open.

Importantly, I left open the question of how to disjoin the various sequence forms that a function can take. In the case of 'map', 'filter' and 'fold', the resulting forms can be complex. See discussion inline where I suggest an option, but ask for feedback.

Thanks!

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs]: this comment has been solved indirectly by the refactor to use "sequence_A, A" style types

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs]: The new version uses copy/pasted strings to ensure consistency.

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs]: I updated the docs to use a kind of "virtual type", where a sequence of type "X" is said to have type "sequence_X". 

Then, each description using such a sequence type has the quote
"Applicable sequence types are `(list A)`, `buff`, `string-ascii` and `string-utf8`,
for which the corresponding element types are, repsectively, `A`, `(buff 1)`, `(string-ascii 1)` and `(string-utf8 1)`."

**Comment by gregorycoppola** [comment]
I just pushed a version of this change that refactors the sequence arguments for readability and consistency.

See comment https://github.com/blockstack/stacks-blockchain/pull/2676#discussion_r646839569

File: src/vm/docs/mod.rs
Comment (pavitthrap) [reply src/vm/docs/mod.rs]: Ya, I don't think there was a right way to do this on second thought

File: src/vm/docs/mod.rs
Comment (pavitthrap) [inline src/vm/docs/mod.rs:467]: output type here can be `(list X)` right?

File: src/vm/docs/mod.rs
Comment (pavitthrap) [inline src/vm/docs/mod.rs:594]: output type is still an optional here?

**Review by pavitthrap — APPROVED** [review]
Just had a few minor comments left! Thanks for updating the docs :)

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:467]: good catch!

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:594]: thanks for catching

File: src/vm/docs/mod.rs
Comment (kantai) [inline src/vm/docs/mod.rs:492]: repsectively -> respectively

File: src/vm/docs/mod.rs
Comment (kantai) [inline src/vm/docs/mod.rs:472]: repsectively -> respectively

File: src/vm/docs/mod.rs
Comment (kantai) [inline src/vm/docs/mod.rs:519]: repsectively -> respectively

**Review by kantai — APPROVED** [review]
LGTM, just a couple typos!

File: src/vm/docs/mod.rs
Comment (kantai) [reply src/vm/docs/mod.rs]: To me, it is much more readable now with the `sequence` virtual type.

**Review by jcnelson — APPROVED** [review]
Thanks @gregorycoppola!  LGTM.  Please make sure all tests pass before merging -- looks like there's a unit test failure?

**Comment by psq** [comment]
After all these iterations, the final version is much better now.  Thank you!

**Pull request merged**
